base/dict.jl

```diff
@@ -866,6 +887,162 @@ length(t::ImmutableDict) = count(Returns(true), t)
 empty(::ImmutableDict, ::Type{K}, ::Type{V}) where {K, V} = ImmutableDict{K,V}()
 
+function setindex(d::ImmutableDict{K,V}, v::V, k::K) where {K,V}
+   if isdefined(d, :parent)
+        if isequal(d.key, k)
+            d0 = d.parent
+            v0 = v
+            k0 = k
+        else
+            d0 = setindex(d.parent, v, k)
+            v0 = d.value
+            k0 = d.key
+        end
+    else
+        d0 = d
+        v0 = v
+        k0 = k
+    end
+    ImmutableDict{K,V}(d0, k0, v0)
+end
+
+function delete(d::ImmutableDict{K,V}, key) where {K,V}
+    if isdefined(d, :parent)
+        if isequal(d.key, key)
+            d.parent
+        else
+            ImmutableDict{K,V}(delete(d.parent, key), d.key, d.value)
+        end
+    else
+        d
+    end
+end
+
 _similar_for(c::AbstractDict, ::Type{Pair{K,V}}, itr, isz, len) where {K, V} = empty(c, K, V)
 _similar_for(c::AbstractDict, ::Type{T}, itr, isz, len) where {T} =
     throw(ArgumentError("for AbstractDicts, similar requires an element type of Pair;\n  if calling map, consider a comprehension instead"))
+
+
+include("hamt.jl")
+using .HashArrayMappedTries
+const HAMT = HashArrayMappedTries
+
+"""
+    PersistentDict
+
+A space efficient persistent dictionary. It based on an hash array mapped trie (HAMT).
+
+Persistent operations are [`Base.setindex`](@ref) which returns a new dictionary with `k=>v` inserted,
+and [`Base.delete`](@ref) which returns a new dictionary with the mapping for `k` removed.
+
+It is space efficient since HAMT's support structural shareing through path-copying.
+"""
+struct PersistentDict{K,V} <: AbstractDict{K,V}
+    trie::HAMT.HAMT{K,V}
+end
+PersistentDict{K,V}() where {K,V} = PersistentDict(HAMT.HAMT{K,V}())
+# TODO: Add HAMT with single element
+PersistentDict(KV::Pair{K,V}) where {K,V} = setindex(PersistentDict{K,V}(), KV.second, KV.first)
+PersistentDict(dict::PersistentDict{K,V}, KV::Pair{K,V}) where {K,V} = setindex(dict, KV.second, KV.first)
+
+function PersistentDict(kv::Pair, rest::Pair...)
+    dict = PersistentDict(kv)
+    for kv in rest
+        key, value = kv
+        dict = setindex(dict, value, key)
+    end
+    return dict
+end
+
+eltype(::PersistentDict{K,V}) where {K,V} = Pair{K,V}
+
+function in(key_val::Pair{K,V}, dict::PersistentDict{K,V}, valcmp=(==)) where {K,V}
+    trie = dict.trie
+    if HAMT.islevel_empty(trie)
+        return false
+    end
+
+    key, val = key_val
+
+    h = hash(key)
+    found, present, trie, i, _, _, _ = HAMT.path(trie, key, h)
+    if found && present
+        leaf = @inbounds trie.data[i]::HAMT.Leaf{K,V}
+        return valcmp(val, leaf.val) && return true
+    end
+    return false
+end
+
+function haskey(dict::PersistentDict{K}, key::K) where K
+    trie = dict.trie
+    h = hash(key)
+    found, present, _, _, _, _, _ = HAMT.path(trie, key, h)
+    return found && present
+end
+
+function getindex(dict::PersistentDict{K,V}, key::K) where {K,V}
```

Comment:
Do we want to have separate methods for these when the keys can be anything (`getindex(::PersistentDict{Any}, @nospecialize(key))`)?